src/hotspot/share/code/nmethod.hpp

```diff
@@ -188,8 +188,23 @@ class nmethod : public CodeBlob {
 
   Method*   _method;
 
-  // To support simple linked-list chaining of nmethods:
-  nmethod*  _osr_link;         // from InstanceKlass::osr_nmethods_head
+  // To reduce header size union fields which usages do not overlap.
+  union {
+    // To support simple linked-list chaining of nmethods:
+    nmethod*  _osr_link; // from InstanceKlass::osr_nmethods_head
+    struct {
+      // These are used for compiled synchronized native methods to
+      // locate the owner and stack slot for the BasicLock. They are
+      // needed because there is no debug information for compiled native
+      // wrappers and the oop maps are insufficient to allow
+      // frame::retrieve_receiver() to work. Currently they are expected
+      // to be byte offsets from the Java stack pointer for maximum code
+      // sharing between platforms. JVMTI's GetLocalInstance() uses these
+      // offsets to find the receiver for non-static native wrapper frames.
+      ByteSize _native_receiver_sp_offset;
+      ByteSize _native_basic_lock_sp_offset;
```

Comment:
Don't we need an assert in the accessor functions to make sure nmethod is native or not?

Reply:
I thought about that but in both places where these accessors are called (`frame::get_native_monitor()` and `frame::get_native_receiver()`) there are such asserts already:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/frame.cpp#L1085

Reply:
OK, but I'd rather see it in the accessors too.  Some users are checking for method()->is_native() and others are checking for is_osr_method(), so we need to make sure those are always mutually exclusive: method()->is_native() != is_osr_method().

Reply:
We have separate `nmethod()` constructor for native method wrappers and it sets `_entry_bci = InvocationEntryBci;`. So it is impossible to have OSR native method wrapper.
But I agree with adding assert into accessors to catch accidental usage for not native methods.